src/Stratis.SmartContracts.CLR.Tests/SmartContracts/SystemContractsDictionary.cs

```diff
@@ -0,0 +1,205 @@
+﻿using Stratis.SmartContracts;
+using ECRecover=Stratis.SCL.Crypto.ECRecover;
+
+public struct WhiteListEntry
+{
+    public UInt256 CodeHash;
+    public Address LastAddress;
+    public string Name;
+}
+
+public class SystemContractsDictionary : SmartContract
+{
+    const string primaryGroup = "main";
+
+    public SystemContractsDictionary(ISmartContractState state) : base(state)
+    {
+        this.State.SetArray($"Signatories:{primaryGroup}", new[] { new Address(0, 0, 0, 0, 0), new Address(0, 0, 0, 0, 1), new Address(0, 0, 0, 0, 2) });
+        this.State.SetUInt32($"Quorum:{primaryGroup}", 2);
+    }
+
+    public Address[] Signatories => GetSignatories(primaryGroup);
+
+    public uint Quorum => GetQuorum(primaryGroup);
+
+    private void VerifySignatures(byte[] signatures, string authorizationChallenge)
+    {
+        string[] sigs = this.Serializer.ToArray<string>(signatures);
+        Assert(ECRecover.TryVerifySignatures(sigs, System.Text.Encoding.ASCII.GetBytes(authorizationChallenge), this.Signatories, out Address[] verifiedSignatures) && verifiedSignatures.Length >= this.Quorum,
+            $"Please provide {this.Quorum} valid signatures for '{authorizationChallenge}'.");
+    }
+
+    public Address[] GetSignatories(string group)
+    {
+        Assert(!string.IsNullOrEmpty(group));
+        return this.State.GetArray<Address>($"Signatories:{group}");
+    }
+
+    public uint GetQuorum(string group)
+    {
+        Assert(!string.IsNullOrEmpty(group));
+        return this.State.GetUInt32($"Quorum:{group}");
+    }
+
+    public void AddSignatory(byte[] signatures, string group, Address address, uint newSize, uint newQuorum)
+    {
+        Assert(!string.IsNullOrEmpty(group));
+        Assert(newSize >= newQuorum, "The number of signatories can't be less than the quorum.");
+
+        Address[] signatories = this.GetSignatories(group);
+        for (int i = 0; i < signatories.Length; i++)
+            Assert(signatories[i] != address, "The signatory already exists.");
+
+        Assert((signatories.Length + 1) == newSize, "The expected size is incorrect.");
+
+        uint nonce = this.State.GetUInt32($"GroupNonce:{group}");
+
+        this.VerifySignatures(signatures, $"{nameof(AddSignatory)}(Nonce:{nonce},Group:{group},Address:{address},NewSize:{newSize},NewQuorum:{newQuorum})");
+
+        System.Array.Resize(ref signatories, signatories.Length + 1);
+        signatories[signatories.Length - 1] = address;
+
+        this.State.SetArray($"Signatories:{group}", signatories);
+        this.State.SetUInt32($"Quorum:{group}", newQuorum);
+        this.State.SetUInt32($"GroupNonce:{group}", nonce + 1);
+    }
+
+    public void RemoveSignatory(byte[] signatures, string group, Address address, uint newSize, uint newQuorum)
+    {
```

Comment:
These methods can be called by anyone ?

Reply:
Yes and incur gas costs. We do perform the signature check before making any updates.

Reply:
It is really needed to specify newSize and newQuorum, why not to automate it ?

Reply:
I really want the user to be very exact about the change being accomplished especially due to how important the quorum count is in relation to the signatory count. The user must know exactly what is being accomplished in that regards and show that knowledge by specifying the values explicitly.

Reply:
Hımm but the transaction is going to fail and cost fee if anyone going to call these methods in same block or in prev blocks.

Reply:
Not sure I understand what you're saying. Will mull it over.